### PR TITLE
(core): Add bulk method

### PR DESCRIPTION
**Pull request by skarekrow**
This allows one call to a middleware method to be iterated on with multiple items.

Example:
midclt call core.bulk jail.do_delete '["foo", "bar", "baz"]'
midclt call core.bulk jail.fetch '[{"release": "11.0-RELEASE"}, {"release": "11.1-RELEASE"}]'

This is lacking exception handling for Jobs

### CODE REVIEW

File: src/middlewared/middlewared/plugins/jail.py [hunk 1]
```diff
@@ -142,6 +142,7 @@ def check_dataset_existence(self):
     @private
     def check_jail_existence(self, jail, skip=True):
         """Wrapper for iocage's API, as a few commands aren't ported to it"""
+        print(jail)
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: ?

File: src/middlewared/middlewared/plugins/jail.py [hunk 2]
```diff
@@ -153,7 +154,7 @@ def check_jail_existence(self, jail, skip=True):
     @accepts(
         Dict("options",
              Str("release"),
-             Str("server", default="ftp.freebsd.org"),
+             Str("server", default="download.freebsd.org"),
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: Lets do it in another commit

File: src/middlewared/middlewared/service.py [hunk 3]
```diff
@@ -10,20 +8,25 @@
...
+PeriodicTaskDescriptor = namedtuple(
+    "PeriodicTaskDescriptor", ["interval", "run_on_start"])
 
 
 def item_method(fn):
     """Flag method as an item method.
     That means it operates over a single item in the collection,
     by an unique identifier."""
     fn._item_method = True
+
```
Comment (william-gr) [inline src/middlewared/middlewared/service.py]: Can we stop doing that on existing code? :P

File: src/middlewared/middlewared/service.py [hunk 4]
```diff
@@ -183,7 +205,8 @@ class SystemServiceService(ConfigService):
                                    {'prefix': self._config.datastore_prefix})
 
         enabled = (await self.middleware.call(
-            'datastore.query', 'services.services', [('srv_service', '=', self._config.service)], {'get': True}
```
Comment (william-gr) [inline src/middlewared/middlewared/service.py]: Can we stop doing that on existing code as well? We are not following 80 char limit here

This is twicthing my OCD haha

File: src/middlewared/middlewared/service.py [hunk 5]
```diff
@@ -496,23 +558,59 @@ def sleep_fn():
...
+        current_progress = 0
+
+        for p in params:
+            try:
+                msg = await self.middleware.call(method, p)
+
+                if isinstance(msg, Job):
+                    # msg is a job right now
+                    # TODO: Doesn't get job exceptions
+                    msg = await msg.wait()
```
Comment (william-gr) [inline src/middlewared/middlewared/service.py]:
```
diff --git a/src/middlewared/middlewared/service.py b/src/middlewared/middlewared/service.py
index 899ac622a2..c5521b5c5b 100644
--- a/src/middlewared/middlewared/service.py
+++ b/src/middlewared/middlewared/service.py
@@ -600,15 +600,20 @@ class CoreService(Service):
         for p in params:
             try:
                 msg = await self.middleware.call(method, p)
+                error = None
 
                 if isinstance(msg, Job):
                     # msg is a job right now
                     # TODO: Doesn't get job exceptions
+                    job = msg
                     msg = await msg.wait()
 
-                statuses.append({"result": msg, "exception": None})
+                    if job.error:
+                        error = job.error
+
+                statuses.append({"result": msg, "error": error})
             except Exception as e:
-                statuses.append({"result": str(e), "exception": e})
+                statuses.append({"result": None, "error": str(e)})
 
             current_progress += progress_step
             job.set_progress(current_progress)

```

File: src/middlewared/middlewared/service.py [hunk 6]
```diff
@@ -516,3 +516,39 @@ def sleep_fn():
             import pydevd
             pydevd.stoptrace()
             pydevd.settrace(host=options['host'])
+
+    @accepts(Str("method"), List("params"))
+    @job(lock=lambda args: f"bulk:{args[-1]}")
```
Comment (william-gr) [inline src/middlewared/middlewared/service.py]: `args[-1]` will lock based on `params`. We probably want `args[0]` which is `method`.

**Comment by william-gr** [comment]
@themylogin  can you review this as well please? I would like another pair of eyes

Thanks!

File: src/middlewared/middlewared/service.py (hunk 6 above)
Comment (themylogin) [reply src/middlewared/middlewared/service.py]: Why do we want to lock at all? Won't `method` do locking itself if it's needed?

**Comment by skarekrow** [comment]
Sorry @themylogin had to rebase this thanks to my vim being overly pushy :)

As for the locking comment, there are a lot of methods that don't lock via jobs. This ensures we have a lock no matter what.

**Comment by skarekrow** [comment]
Jenkins: retest this please
